File: src/DashSpec.Core/Parsing/ModuleIncludeState.cs
using DashSpec.Core.Model;

namespace DashSpec.Core.Parsing;

internal enum DocumentModuleKind
{
    Dashboard,
    Tab,
}

internal sealed class ModuleIncludeState
{
    private readonly Dictionary<string, SpecIncludeFragment> _diagrams =
        new(StringComparer.OrdinalIgnoreCase);

    private readonly Dictionary<string, PresentationBlock> _chartChromePresets =
        new(StringComparer.OrdinalIgnoreCase);

    private readonly Dictionary<string, TooltipDefinition> _tooltips =
        new(StringComparer.OrdinalIgnoreCase);

    public LayoutBoardDefinition? LayoutBoard { get; private set; }

    public LayoutBoardDefinition? ToolbarBoard { get; private set; }

    public bool TryGetDiagram(string id, out SpecIncludeFragment fragment) =>
        _diagrams.TryGetValue(id, out fragment!);

    public void RegisterDiagram(string id, SpecIncludeFragment fragment)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(id);
        if (_diagrams.ContainsKey(id))
        {
            throw new DashSpecParseException($"Duplicate diagram id '{id}' in module includes.");
        }

        _diagrams[id] = fragment;
    }

    public void RegisterTooltip(string id, TooltipDefinition definition)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(id);
        ArgumentNullException.ThrowIfNull(definition);
        if (_tooltips.ContainsKey(id))
        {
            throw new DashSpecParseException($"Duplicate tooltip id '{id}' in module includes.");
        }

        _tooltips[id] = definition;
    }

    public void RegisterChartChromePreset(string id, PresentationBlock block)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(id);
        ArgumentNullException.ThrowIfNull(block);
        if (_chartChromePresets.ContainsKey(id))
        {
            throw new DashSpecParseException($"Duplicate chart chrome preset '{id}' in module includes.");
        }

        _chartChromePresets[id] = block;
    }

    public IReadOnlyDictionary<string, PresentationBlock> ExportChartChromePresets()
    {
        if (_chartChromePresets.Count == 0)
        {
            return DashboardDocument.EmptyModuleChartChromePresets;
        }

        return new Dictionary<string, PresentationBlock>(_chartChromePresets, StringComparer.OrdinalIgnoreCase);
    }

    public IReadOnlyDictionary<string, TooltipDefinition> ExportTooltips()
    {
        if (_tooltips.Count == 0)
        {
            return DashboardDocument.EmptyModuleTooltips;
        }

        return new Dictionary<string, TooltipDefinition>(_tooltips, StringComparer.OrdinalIgnoreCase);
    }

    public IReadOnlyDictionary<string, ModuleDiagramDefinition> ExportDefinitions()
    {
        if (_diagrams.Count == 0)
        {
            return DashboardDocument.EmptyModuleDiagrams;
        }

        return _diagrams.ToDictionary(
            static kv => kv.Key,
            kv =>
            {
                var fragment = kv.Value;
                var tooltips = MergeTooltipMaps(_tooltips, fragment.Tooltips);
                TooltipDefinition? tooltip = null;
                if (fragment.Inspect?.TooltipId is { } tooltipId &&
                    tooltips.TryGetValue(tooltipId, out var resolved))
                {
                    tooltip = resolved;
                }

                return new ModuleDiagramDefinition(
                    fragment.Diagram ?? throw new DashSpecParseException($"Diagram include '{kv.Key}' has no kind block."),
                    fragment.Presentation,
                    fragment.SeriesTransform,
                    fragment.Inspect,
                    tooltip);
            },
            StringComparer.OrdinalIgnoreCase);
    }

    public void AssignLayoutBoard(LayoutBoardDefinition board, string context)
    {
        if (LayoutBoard is not null)
        {
            throw new DashSpecParseException($"{context} declares more than one card layout board.");
        }

        LayoutModuleScopeValidator.EnsureMatchesIncludeSite(board, LayoutScope.Tab, context);
        LayoutBoard = board;
    }

    public void AssignToolbarBoard(LayoutBoardDefinition board, string context)
    {
        if (ToolbarBoard is not null)
        {
            throw new DashSpecParseException($"{context} declares more than one toolbar layout board.");
        }

        LayoutModuleScopeValidator.EnsureMatchesIncludeSite(board, LayoutScope.Toolbar, context);
        ToolbarBoard = board;
    }

    private static Dictionary<string, TooltipDefinition> MergeTooltipMaps(
        IReadOnlyDictionary<string, TooltipDefinition> stateTooltips,
        IReadOnlyDictionary<string, TooltipDefinition>? fragmentTooltips)
    {
        var merged = new Dictionary<string, TooltipDefinition>(stateTooltips, StringComparer.OrdinalIgnoreCase);
        if (fragmentTooltips is null)
        {
            return merged;
        }

        foreach (var (key, value) in fragmentTooltips)
        {
            merged[key] = value;
        }

        return merged;
    }
}

File: src/DashSpec.Modeling.Parse/Include/ModuleIncludeState.fs
namespace DashSpec.Modeling.Parse.Include

open System
open System.Collections.Generic
open DashSpec.Modeling.Core
open DashSpec.Modeling.Parse.Diagram
open DashSpec.Modeling.Parse.Layout
open DashSpec.Modeling.Parse.Presentation
open DashSpec.Modeling.Parse.Tooltip

type ModuleIncludeState() =
    let diagrams = Dictionary<string, SpecIncludeFragment>(StringComparer.OrdinalIgnoreCase)
    let chartChromePresets = Dictionary<string, PresentationBlock>(StringComparer.OrdinalIgnoreCase)
    let tooltips = Dictionary<string, TooltipDefinition>(StringComparer.OrdinalIgnoreCase)

    let mutable layoutBoard: LayoutBoardDefinition option = None
    let mutable toolbarBoard: LayoutBoardDefinition option = None

    member _.LayoutBoard = layoutBoard
    member _.ToolbarBoard = toolbarBoard

    member _.TryGetDiagram(id: string, fragment: outref<SpecIncludeFragment>) =
        match diagrams.TryGetValue id with
        | true, value ->
            fragment <- value
            true
        | false, _ -> false

    member this.RegisterDiagram(id: string, fragment: SpecIncludeFragment) =
        if String.IsNullOrWhiteSpace id then invalidArg "id" "Diagram id is required."
        if diagrams.ContainsKey id then
            raise (DashSpecParseException($"Duplicate diagram id '{id}' in module includes."))
        diagrams.[id] <- fragment

    member this.RegisterTooltip(id: string, definition: TooltipDefinition) =
        if String.IsNullOrWhiteSpace id then invalidArg "id" "Tooltip id is required."
        if tooltips.ContainsKey id then
            raise (DashSpecParseException($"Duplicate tooltip id '{id}' in module includes."))
        tooltips.[id] <- definition

    member this.RegisterChartChromePreset(id: string, block: PresentationBlock) =
        if String.IsNullOrWhiteSpace id then invalidArg "id" "Preset id is required."
        if chartChromePresets.ContainsKey id then
            raise (DashSpecParseException($"Duplicate chart chrome preset '{id}' in module includes."))
        chartChromePresets.[id] <- block

    member this.AssignLayoutBoard(board: LayoutBoardDefinition, context: string) =
        if layoutBoard.IsSome then
            raise (DashSpecParseException($"{context} declares more than one card layout board."))
        LayoutModuleScopeValidator.ensureMatchesIncludeSite board LayoutScope.Tab context
        layoutBoard <- Some board

    member this.AssignToolbarBoard(board: LayoutBoardDefinition, context: string) =
        if toolbarBoard.IsSome then
            raise (DashSpecParseException($"{context} declares more than one toolbar layout board."))
        LayoutModuleScopeValidator.ensureMatchesIncludeSite board LayoutScope.Toolbar context
        toolbarBoard <- Some board

    member _.ExportChartChromePresets() = chartChromePresets :> IReadOnlyDictionary<_, _>
    member _.ExportTooltips() = tooltips :> IReadOnlyDictionary<_, _>
    member _.Diagrams = diagrams :> IReadOnlyDictionary<_, _>
